CI on pull request fails to detect package-lock.json version error

### Description
When working on https://github.com/atom/atom/pull/19130, I ran into a situation where all CI checks passed on the pull request, leading me to believe that everything was fine with changes. However, when I merged the pull request, one of the CI checks failed.

On the pull request, CI ran two checks:

- ✅ [**Azure DevOps** - *Atom Pull Requests* Pipeline](https://github.visualstudio.com/Atom/_build/results?buildId=37525)
- ✅ [**AppVeyor **- *continuous-integration/appveyor/pr * context](https://ci.appveyor.com/project/Atom/atom/builds/23745368)

After merging the pull request, CI ran three checks:

- ✅ **Azure DevOps** - *Atom Production Branches* Pipeline
- ✅ **AppVeyor** - *continuous-integration/appveyor/branch* context
- ❌ [**Travis CI** - *continuous-integration/travis-ci/push* context](https://travis-ci.org/atom/atom/builds/518371554#L737)

The Travis CI check identified a legitimate issue, but CI should have alerted us to that issue before the pull request was merged:

> `npm ERR! Invalid: lock file's fuzzy-finder@https://www.atom.io/api/packages/fuzzy-finder/versions/1.12.1/tarball does not satisfy fuzzy-finder@https://www.atom.io/api/packages/fuzzy-finder/versions/1.13.0/tarball`

### Screenshots

<img width="772" alt="Screen Shot 2019-04-12 at 11 50 56 AM" src="https://user-images.githubusercontent.com/2988/56050407-d45de680-5d19-11e9-9e8d-a6a78e0b3838.png">

CI checks on pull request:

<img width="776" alt="Screen Shot 2019-04-12 at 11 51 11 AM" src="https://user-images.githubusercontent.com/2988/56050412-dcb62180-5d19-11e9-8274-8d597c2435ce.png">

CI checks after merging:

<img width="784" alt="Screen Shot 2019-04-12 at 11 51 53 AM" src="https://user-images.githubusercontent.com/2988/56050413-dcb62180-5d19-11e9-9df7-ecaa1b7eb3f5.png">

### Expected vs. actual behavior

**Expected behavior:** CI fails when a `package-lock.json` problem is present on the pull request.

**Actual behavior:** CI passes when a `package-lock.json` problem is present on the pull request.

--- a/appveyor.yml
+++ b/appveyor.yml
@@ -20,6 +20,7 @@ environment:
   global:
     ATOM_DEV_RESOURCE_PATH: c:\projects\atom
     ATOM_JASMINE_REPORTER: list
+    CI: true
     TEST_JUNIT_XML_ROOT: c:\projects\junit-test-results
     NODE_VERSION: 8.9.3
 

--- a/script/bootstrap
+++ b/script/bootstrap
@@ -26,7 +26,7 @@ if (!ci && process.env.CI === 'true' && process.argv.indexOf('--no-ci') === -1) 
   ci = true
 }
 
-verifyMachineRequirements()
+verifyMachineRequirements(ci)
 
 if (dependenciesFingerprint.isOutdated()) {
   cleanDependencies()

--- a/script/config.js
+++ b/script/config.js
@@ -77,6 +77,8 @@ function getApmBinPath () {
 }
 
 function getNpmBinPath (external = false) {
+  if (process.env.NPM_BIN_PATH) return process.env.NPM_BIN_PATH
+
   const npmBinName = process.platform === 'win32' ? 'npm.cmd' : 'npm'
   const localNpmBinPath = path.resolve(repositoryRootPath, 'script', 'node_modules', '.bin', npmBinName)
   return !external && fs.existsSync(localNpmBinPath) ? localNpmBinPath : npmBinName

--- a/script/lib/verify-machine-requirements.js
+++ b/script/lib/verify-machine-requirements.js
@@ -6,9 +6,9 @@ const path = require('path')
 
 const CONFIG = require('../config')
 
-module.exports = function () {
+module.exports = function (ci) {
   verifyNode()
-  verifyNpm()
+  verifyNpm(ci)
   if (process.platform === 'win32') {
     verifyPython()
   }
@@ -27,14 +27,15 @@ function verifyNode () {
   }
 }
 
-function verifyNpm () {
-  const stdout = childProcess.execFileSync(CONFIG.getNpmBinPath(), ['--version'], {env: process.env})
+function verifyNpm (ci) {
+  const stdout = childProcess.execFileSync(CONFIG.getNpmBinPath(ci), ['--version'], {env: process.env})
   const fullVersion = stdout.toString().trim()
   const majorVersion = fullVersion.split('.')[0]
-  if (majorVersion >= 3) {
+  const oldestMajorVersionSupported = ci ? 6 : 3
+  if (majorVersion >= oldestMajorVersionSupported) {
     console.log(`Npm:\tv${fullVersion}`)
   } else {
-    throw new Error(`npm v3+ is required to build Atom. npm v${fullVersion} was detected.`)
+    throw new Error(`npm v${oldestMajorVersionSupported}+ is required to build Atom. npm v${fullVersion} was detected.`)
   }
 }
 

--- a/script/vsts/platforms/linux.yml
+++ b/script/vsts/platforms/linux.yml
@@ -14,6 +14,9 @@ jobs:
   steps:
   - script: script/bootstrap
     displayName: Bootstrap build environment
+    env:
+      CI: true
+      CI_PROVIDER: VSTS
 
   - script: script/lint
     displayName: Run linter

--- a/script/vsts/platforms/macos.yml
+++ b/script/vsts/platforms/macos.yml
@@ -19,7 +19,7 @@ jobs:
   - script: npm install --global npm@6.2.0
     displayName: Update npm
 
-  - script: script/bootstrap
+  - script: export PATH="/usr/local/bin:${PATH}" && script/bootstrap
     displayName: Bootstrap build environment
     env:
       CI: true

--- a/script/vsts/platforms/macos.yml
+++ b/script/vsts/platforms/macos.yml
@@ -19,11 +19,12 @@ jobs:
   - script: npm install --global npm@6.2.0
     displayName: Update npm
 
-  - script: export PATH="/usr/local/bin:${PATH}" && script/bootstrap
+  - script: script/bootstrap
     displayName: Bootstrap build environment
     env:
       CI: true
       CI_PROVIDER: VSTS
+      NPM_BIN_PATH: /usr/local/bin/npm
 
   - script: script/lint
     displayName: Run linter

--- a/script/vsts/platforms/macos.yml
+++ b/script/vsts/platforms/macos.yml
@@ -21,6 +21,9 @@ jobs:
 
   - script: script/bootstrap
     displayName: Bootstrap build environment
+    env:
+      CI: true
+      CI_PROVIDER: VSTS
 
   - script: script/lint
     displayName: Run linter

--- a/script/vsts/platforms/windows.yml
+++ b/script/vsts/platforms/windows.yml
@@ -40,6 +40,8 @@ jobs:
       node script\vsts\windows-run.js script\bootstrap.cmd
     env:
       BUILD_ARCH: $(buildArch)
+      CI: true
+      CI_PROVIDER: VSTS
     displayName: Bootstrap build environment
 
   - script: node script\vsts\windows-run.js script\lint.cmd

--- a/script/vsts/platforms/windows.yml
+++ b/script/vsts/platforms/windows.yml
@@ -27,6 +27,9 @@ jobs:
   - script: |
       ECHO Installing npm-windows-upgrade
       npm install --global --production npm-windows-upgrade
+    displayName: Install npm-windows-upgrade
+
+  - script: |
       ECHO Upgrading npm
       npm-windows-upgrade --no-spinner --no-prompt --npm-version 6.2.0
     displayName: Install npm 6.2.0
@@ -42,6 +45,7 @@ jobs:
       BUILD_ARCH: $(buildArch)
       CI: true
       CI_PROVIDER: VSTS
+      NPM_BIN_PATH: "D:\\a\\_tool\\node\\8.9.3\\x64\\npm.cmd"
     displayName: Bootstrap build environment
 
   - script: node script\vsts\windows-run.js script\lint.cmd

